src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,801 @@
+    mState[key] = ler;
+
+    if (entry.data.type() == OFFER)
+    {
+        OfferEntry const& offer = entry.data.offer();
+        invalidateLoadBestOfferContext(offer.selling, offer.buying);
+    }
+
+    return ler;
+}
+
+LedgerState::StateEntry
+LedgerState::createHelper(LedgerEntry const& entry, LedgerKey const& key)
+{
+    StateEntry ler;
+
+    auto iter = mState.find(key);
+    if (iter != mState.end())
+    {
+        if (iter->second->ignoreInvalid().entry())
+        {
+            throw std::runtime_error("Key already exists in memory");
+        }
+        auto newEntry = std::make_shared<LedgerEntry>(entry);
+        ler = StateEntry(new LedgerEntryReference(newEntry, nullptr));
+    }
+    else if (mParent)
+    {
+        ler = mParent->createHelper(entry, key);
+    }
+    else
+    {
+        if (loadFromDatabase(key))
+        {
+            throw std::runtime_error("Key already exists in database");
+        }
+        auto newEntry = std::make_shared<LedgerEntry>(entry);
+        ler = StateEntry(new LedgerEntryReference(newEntry, nullptr));
+    }
+    return ler;
+}
+
+LedgerState::StateEntry
+LedgerState::load(LedgerKey const& key)
+{
+    assert(mParent || mRoot);
+    assert(!mChild);
+
+    auto ler = loadHelper(key);
+    mState[key] = ler;
+
+    if (ler->entry()->data.type() == OFFER)
+    {
+        OfferEntry const& offer = ler->entry()->data.offer();
+        invalidateLoadBestOfferContext(offer.selling, offer.buying);
+    }
+
+    return ler;
+}
+
+LedgerState::StateEntry
+LedgerState::loadHelper(LedgerKey const& key)
+{
+    StateEntry ler;
+
+    auto iter = mState.find(key);
+    if (iter != mState.end())
+    {
+        if (iter->second->valid())
+        {
+            throw std::runtime_error(
+                "Valid LedgerEntryReference already exists for this key");
+        }
+
+        auto const& entry = iter->second->ignoreInvalid().entry();
+        if (!entry)
+        {
+            throw std::runtime_error(
+                "Key exists in memory but LedgerEntry has been erased");
+        }
+
+        if (!mChild)
+        {
+            auto const& previous =
+                iter->second->ignoreInvalid().previousEntry();
+            ler = StateEntry(new LedgerEntryReference(entry, previous));
+        }
+        else
+        {
+            ler = StateEntry(new LedgerEntryReference(entry, entry));
+        }
+    }
+    else if (mParent)
+    {
+        ler = mParent->loadHelper(key);
+    }
+    else
+    {
+        ler = loadFromDatabase(key);
+        if (!ler)
+        {
+            throw std::runtime_error("Key does not exist in database");
+        }
+    }
+    return ler;
+}
+
+LedgerState::StateHeader
+LedgerState::loadHeader()
+{
+    assert(mParent || mRoot);
+    assert(!mChild);
+
+    mHeader = loadHeaderHelper();
+    auto& header = mHeader->header();
+    ++header.ledgerSeq;
+    header.previousLedgerHash = sha256(xdr::xdr_to_opaque(header));
```

Comment:
just wondering, will it use proper value of ledgerSeq as it was just incremented? seems strange to be that methods named "loadHeader" does modifications on header object

Reply:
Actually, this doesn't look quite right to me. I'm going to revisit this now. I hadn't written any tests for this function yet because I wasn't sure exactly what I wanted to test, but I'm going to write one now.

Reply:
Fixed.